pkg/datapath/loader/compile.go

```diff
@@ -0,0 +1,182 @@
+// Copyright 2017-2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package loader
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"runtime"
+)
+
+// OutputType determines the type to be generated by the compilation steps.
+type OutputType string
+
+const (
+	outputObject   = OutputType("obj")
+	outputAssembly = OutputType("asm")
+	outputSource   = OutputType("c")
+)
+
+// progInfo describes a program to be compiled with the expected output format
+type progInfo struct {
+	// Source is the program source (base) filename to be compiled
+	Source string
+	// Output is the expected (base) filename produced from the source
+	Output string
+	// OutputType to be created by LLVM
+	OutputType OutputType
+}
+
+// directoryInfo includes relevant directories for compilation and linking
+type directoryInfo struct {
+	// Library contains the library code to be used for compilation
+	Library string
+	// Runtime contains headers for compilation
+	Runtime string
+	// Output is the directory where the files will be stored
+	Output string
+}
+
+var (
+	compiler       = "clang"
+	linker         = "llc"
+	standardCFlags = []string{"-O2", "-target", "bpf",
+		fmt.Sprintf("-D__NR_CPUS__=%d", runtime.NumCPU()),
+		"-Wno-address-of-packed-member", "-Wno-unknown-warning-option"}
+	standardLDFlags = []string{"-march=bpf", "-mcpu=probe"}
+
+	endpointPrefix   = "bpf_lxc"
+	endpointProg     = fmt.Sprintf("%s.%s", endpointPrefix, outputSource)
+	endpointObj      = fmt.Sprintf("%s.%s", endpointPrefix, outputObject)
+	endpointObjDebug = fmt.Sprintf("%s.dbg.%s", endpointPrefix, outputObject)
+	endpointAsm      = fmt.Sprintf("%s.%s", endpointPrefix, outputAssembly)
+
+	debugProgs = []*progInfo{
+		{
+			Source:     endpointProg,
+			Output:     endpointObjDebug,
+			OutputType: outputObject,
+		},
+		{
+			Source:     endpointProg,
+			Output:     endpointAsm,
+			OutputType: outputAssembly,
+		},
+		{
+			Source:     endpointProg,
+			Output:     endpointProg,
+			OutputType: outputSource,
+		},
+	}
+	datapathProg = &progInfo{
+		Source:     endpointProg,
+		Output:     endpointObj,
+		OutputType: outputObject,
+	}
+)
+
+// irPath determines a temporary path on the filesystem to store the compiled
+// intermediate representation object for the program.
+//
+// TODO: Remove this and just pipe the 'clang' and 'llc' programs together.
```

Comment:
If you want to have some fun: https://stackoverflow.com/questions/10781516/how-to-pipe-several-commands-in-go